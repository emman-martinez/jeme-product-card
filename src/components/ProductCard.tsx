import React, { createContext, CSSProperties, JSX } from 'react';
import { useProduct } from '../hooks/useProduct';
import {
  InitialValues,
  OnChangeArgs,
  Product,
  ProductCardHandlers,
  ProductContextProps,
} from '../interfaces/interfaces';
import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  children?: (args: ProductCardHandlers) => JSX.Element;
  className?: string;
  initialValues?: InitialValues;
  onChange?: (args: OnChangeArgs) => void;
  product: Product;
  style?: CSSProperties;
  value?: number;
}

const ProductCard = ({
  children,
  className,
  initialValues,
  onChange,
  product,
  style,
  value,
}: Props) => {
  const {
    counter,
    isMaxCountReached,
    maxCount,
    increaseBy,
    reset,
  } = useProduct({
    initialValues,
    onChange,
    product,
    value,
  });

  const classStyleName = `${styles.productCard} ${className}`;

  return (
    <Provider value={{ counter, maxCount, product, increaseBy }}>
      <div className={classStyleName} style={style}>
        {children &&
          children({
            count: counter,
            isMaxCountReached,
            maxCount,
            product,
            increaseBy,
            reset,
          })}
      </div>
    </Provider>
  );
};

export default ProductCard;
