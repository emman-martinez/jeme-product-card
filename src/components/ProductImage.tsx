import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

export interface Props {
  className?: string;
  img?: string;
  style?: CSSProperties;
}

const ProductImage = ({ className, img, style }: Props) => {
  const { product } = useContext(ProductContext);
  let imgToShow: string;

  if (img) {
    imgToShow = img;
  } else if (product.img) {
    imgToShow = product.img;
  } else {
    imgToShow = noImage;
  }

  return (
    <img
      alt="Product"
      className={`${styles.productImg} ${className}`}
      src={imgToShow}
      style={style}
    />
  );
};

export default ProductImage;
