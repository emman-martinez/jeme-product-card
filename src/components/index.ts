import { ProductCardHOCProps } from '../interfaces/interfaces';
import ProductCardHOC from './ProductCard';

import ProductButtons from './ProductButtons';
import ProductImage from './ProductImage';
import ProductTitle from './ProductTitle';

export { default as ProductImage } from './ProductImage';
export { default as ProductTitle } from './ProductTitle';
export { default as ProductButtons } from './ProductButtons';

export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHOC, {
  Title: ProductTitle,
  Image: ProductImage,
  Buttons: ProductButtons,
});
