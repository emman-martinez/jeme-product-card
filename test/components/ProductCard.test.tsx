import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { ProductCard } from '../../src/components';
import { product1 } from '../data/products';

describe('ProductCard', () => {
  test('It should render the component correctly', () => {
    const { asFragment } = render(
      <ProductCard product={product1}>
        {() => <h1>Product Card</h1>}
      </ProductCard>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should increase the counter', () => {
    const { asFragment, container } = render(
      <ProductCard product={product1}>
        {({ count, increaseBy }) => (
          <>
            <h1>Product Card</h1>
            <span>{count}</span>
            <button onClick={() => increaseBy(1)}></button>
          </>
        )}
      </ProductCard>
    );

    expect(asFragment()).toMatchSnapshot();

    const initialCount = container.querySelector('span');
    expect(initialCount?.textContent).toBe('0');

    const button = container.querySelector('button');

    fireEvent.click(button!);

    const updatedCount = container.querySelector('span');
    expect(updatedCount?.textContent).toBe('1');
  });
});
