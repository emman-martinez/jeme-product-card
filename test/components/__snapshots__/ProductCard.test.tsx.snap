// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard It should render the component correctly 1`] = `
<DocumentFragment>
  <div
    class="productCard undefined"
  >
    <h1>
      Product Card
    </h1>
  </div>
</DocumentFragment>
`;

exports[`ProductCard should increase the counter 1`] = `
<DocumentFragment>
  <div
    class="productCard undefined"
  >
    <h1>
      Product Card
    </h1>
    <span>
      0
    </span>
    <button />
  </div>
</DocumentFragment>
`;
